ui/src/services/csrfTokenApi.ts

```diff
@@ -0,0 +1,125 @@
+import { AxiosInstance, InternalAxiosRequestConfig, isAxiosError } from 'axios';
+import { randomId, showToast } from '@prefecthq/prefect-design';
+import { Api, AxiosInstanceSetupHook, PrefectConfig } from '@prefecthq/prefect-ui-library'
+
+export type CsrfToken = {
+    token: string | null;
+    expiration: Date | null;
+    client: string;
+    issued: Date | null;
+}
+
+export type CsrfTokenResponse = {
+    token: string;
+    expiration: Date;
+}
+
+export class CsrfTokenApi extends Api {
+    public csrfToken: CsrfToken = {
+        token: null,
+        expiration: null,
+        client: randomId(),
+        issued: null,
+    };
+    public csrfSupportEnabled = true;
+    private refreshTimeout: ReturnType<typeof setTimeout> | null = null;
+    private ongoingRefresh: Promise<void> | null = null;
+
+    public constructor(apiConfig: PrefectConfig, instanceSetupHook: AxiosInstanceSetupHook | null = null) {
+        super(apiConfig, instanceSetupHook)
+        this.startBackgroundTokenRefresh()
+    }
+
+    public async addCsrfHeaders(config: InternalAxiosRequestConfig) {
+        if (this.csrfSupportEnabled) {
+            await this.refreshCsrfToken()
```

Comment:
nit: Took me a minute to figure out that this wouldn't actually refresh the token for every single request. Might be nice to have an async `getCsrfToken` method that abstracts the "maybe refresh" stuff and then just returns the token. 
``` 
const { token, client } = this.getCsrfToken()

config.headers['Prefect-Csrf-Token'] = token
config.headers['Prefect-Csrf-Client'] = client
```

Reply:
That's a good idea, I'll see what I can do with that.